docs/tutorials/boot-tutorial.md

```diff
@@ -325,38 +399,107 @@ linux_binaries = {
 
 Once, all the artifacts are registered the next step is to launch all gem5 jobs. To do that, add the following lines in your script:
 
+## If Using Celery
+
 ```python
 if __name__ == "__main__":
-    boot_types = ['init', 'systemd']
+    boot_types = ['init']
     num_cpus = ['1', '2', '4', '8']
     cpu_types = ['kvm', 'atomic', 'simple', 'o3']
-    mem_types = ['classic', 'ruby']
-
-    for linux in linuxes:
-        for boot_type in boot_types:
-            for cpu in cpu_types:
-                for num_cpu in num_cpus:
-                    for mem in mem_types:
-                        run = gem5Run.createFSRun(
-                            'linux_boot_tests',
-                            'gem5/build/X86/gem5.opt',
-                            'configs-boot-tests/run_exit.py',
-                            'results/run_exit/vmlinux-{}/boot-exit/{}/{}/{}/{}'.
-                            format(linux, cpu, mem, num_cpu, boot_type),
-                            gem5_binary, gem5_repo, experiments_repo,
-                            os.path.join('linux-stable', 'vmlinux'+'-'+linux),
-                            'disk-image/boot-exit/boot-exit-image/boot-exit',
-                            linux_binaries[linux], disk_image,
-                            cpu, mem, num_cpu, boot_type,
-                            timeout = 12*60*60 #12 hours
-                            )
-                        run_gem5_instance.apply_async((run, os.getcwd()))
+    mem_types = ['MI_example', 'MESI_Two_Level', 'MOESI_CMP_directory']
+
+    def createRun(linux, boot_type, cpu, num_cpu, mem):
+
+        if mem == 'MESI_Two_Level':
+            binary_gem5 = 'gem5/build/X86_MESI_Two_Level/gem5.opt'
+            artifact_gem5 = gem5_binary_MESI_Two_Level
+        elif mem == 'MOESI_CMP_directory':
+            binary_gem5 = 'gem5/build/MOESI_CMP_directory/gem5.opt'
+            artifact_gem5 = gem5_binary_MOESI_CMP_directory
+        else:
+            binary_gem5 = 'gem5/build/X86/gem5.opt'
+            artifact_gem5 = gem5_binary
+
+        return gem5Run.createFSRun(
+            'boot experiments with gem5-20.1',
+            binary_gem5,
+            'configs-boot-tests/run_exit.py',
+            'results/run_exit/vmlinux-{}/boot-exit/{}/{}/{}/{}'.
+                format(linux, cpu, mem, num_cpu, boot_type),
+            artifact_gem5, gem5_repo, experiments_repo,
+            os.path.join('linux-stable', 'vmlinux'+'-'+linux),
+            'disk-image/boot-exit/boot-exit-image/boot-exit',
+            linux_binaries[linux], disk_image,
+            cpu, mem, num_cpu, boot_type,
+            timeout = 10*60*60 #10 hours
+            )
+
+    # For the cross product of tests, create a run object.
+    runs = starmap(createRun, product(linuxes, boot_types, cpu_types, num_cpus, mem_types))
+    # Run all of these experiments in parallel
+    for run in runs:
+        run_gem5_instance.apply_async((run, os.getcwd(),))
```

Comment:
This line is the only difference between celery and multiprocessing, right? Most of this looks the same.

No need to copy the code twice.

You can move "createRun" out of the `__name__=="__main__"` statement to make it more modular.

Reply:
Fixed.